contracts/AelinUpFrontDeal.sol

```diff
@@ -0,0 +1,299 @@
+                    emit PoolWith721(
+                        nftCollectionRules[i].collectionAddress,
+                        nftCollectionRules[i].purchaseAmount,
+                        nftCollectionRules[i].purchaseAmountPerToken
+                    );
+                }
+                hasNftList = true;
+            }
+            // if the first address supports 1155, the entire pool only supports 1155
+            else if (NftCheck.supports1155(nftCollectionRules[0].collectionAddress)) {
+                for (uint256 i = 0; i < nftCollectionRules.length; i++) {
+                    require(NftCheck.supports1155(nftCollectionRules[i].collectionAddress), "can only contain 1155");
+                    nftCollectionDetails[nftCollectionRules[i].collectionAddress] = nftCollectionRules[i];
+
+                    for (uint256 j = 0; j < nftCollectionRules[i].tokenIds.length; j++) {
+                        nftId[nftCollectionRules[i].collectionAddress][nftCollectionRules[i].tokenIds[j]] = true;
+                    }
+                    emit PoolWith1155(
+                        nftCollectionRules[i].collectionAddress,
+                        nftCollectionRules[i].purchaseAmount,
+                        nftCollectionRules[i].purchaseAmountPerToken,
+                        nftCollectionRules[i].tokenIds,
+                        nftCollectionRules[i].minTokensEligible
+                    );
+                }
+                hasNftList = true;
+            } else {
+                revert("collection is not compatible");
+            }
+        }
+
+        // deposit underlying token logic
+        // check if the underlyingDealAmount is more than 0,
+        // if yes, transfer it to this contract and store it in `currentDealTokenTotal` and add it
+        if (_underlyingDealTokenAmount > 0) {
+            currentDealTokenTotal += _underlyingDealTokenAmount;
+            IERC20(_dealData.underlyingDealToken).transferFrom(_sponsor, address(this), _underlyingDealTokenAmount);
+            if (_underlyingDealTokenAmount >= _dealData.underlyingDealTokenTotal) {
+                dealDepositComplete = true;
+                emit DealFullyFunded(address(this), block.timestamp, block.timestamp + _dealData.proRataRedemptionPeriod);
+            } else {
+                emit DepositDealToken(_dealData.underlyingDealToken, _sponsor, _underlyingDealTokenAmount);
+            }
+        }
+    }
+
+    function depositDealTokens(uint256 _underlyingDealTokenAmount) public {
+        UpFrontDeal memory _dealData = dealData;
+        uint256 _currentDealTokenTotal = currentDealTokenTotal;
+
+        require(
+            IERC20(_dealData.underlyingDealToken).balanceOf(msg.sender) >= _underlyingDealTokenAmount,
+            "not enough balance"
+        );
+        require(_currentDealTokenTotal < _dealData.maxDealTotalSupply, "already deposited the total");
```

Comment:
we dont need to track the _currentDealTokenTotal. It is just the balance of the contract that we want to know. Essentially, do we have enough deal tokens for the deal. if someone sends deal tokens to the contract but they are not the sponsor or the counter party its their loss. the contract just needs enough deal tokens in it to fulfill the deal

Reply:
Should we include an onlyHolder modifier to call this and have a withdrawExcess that allows the holder to withdraw any underlying deal tokens over the maxDealTotalSupply?

Reply:
yes we need a withdraw function for the holder to be able to withdraw any excess funds that are left in the pool after the investment window is finalized